File: matlab inpementation/NSGA2_main.m

% NSGA2_main.m - Main script for NSGA-II (educational implementation)
clear; clc; close all;
% Choose problem: 'ZDT1','ZDT3','Kursawe','Fonseca','Schaffer1','Schaffer2','BinhKorn','ChankongHaimes'
problem_name = 'ZDT3';

% Algorithm parameters
N = 50; MaxIt = 100;
pc = 0.9; eta_c = 15;
% get problem details
[nVar, VarMin, VarMax] = get_problem_details(problem_name);
pm = 1/nVar; eta_m = 20;

% initialize population
empty.Position=[]; empty.Objectives=[]; empty.Constraints=[];
empty.ConstraintViolation=0; empty.Rank=0; empty.CrowdingDistance=0;
pop = repmat(empty, N, 1);
for i=1:N
    pop(i).Position = unifrnd(VarMin, VarMax, [1,nVar]);
    [pop(i).Objectives, pop(i).Constraints] = evaluate_objectives(pop(i).Position, problem_name);
    pop(i).ConstraintViolation = sum(max(0, pop(i).Constraints));
end
[pop, F] = non_dominated_sort(pop);
pop = crowding_distance(pop, F);

% main loop
for it=1:MaxIt
    fprintf('Iter %d/%d\n', it, MaxIt);
    offspring = repmat(empty, N, 1);
    idx=1;
    while idx<=N
        p1 = tournament_selection(pop);
        p2 = tournament_selection(pop);
        [c1,c2] = genetic_operators(p1.Position, p2.Position, VarMin, VarMax, pc, eta_c, pm, eta_m);
        offspring(idx).Position = c1;
        if idx+1<=N, offspring(idx+1).Position = c2; end
        idx = idx + 2;
    end
    % evaluate offspring
    for i=1:N
        [offspring(i).Objectives, offspring(i).Constraints] = evaluate_objectives(offspring(i).Position, problem_name);
        offspring(i).ConstraintViolation = sum(max(0, offspring(i).Constraints));
    end
    % combine and select
    combined = [pop; offspring];
    [combined, F] = non_dominated_sort(combined);
    combined = crowding_distance(combined, F);
    % build new population
    newpop = repmat(empty, N, 1); count=0; fidx=1;
    while count < N
        front = F{fidx};
        if count + numel(front) <= N
            for j=1:numel(front), count=count+1; newpop(count)=combined(front(j)); end
        else
            cds = arrayfun(@(x) combined(x).CrowdingDistance, front);
            [~, order] = sort(cds, 'descend');
            need = N - count;
            for j=1:need, idx_sel = front(order(j)); count = count+1; newpop(count)=combined(idx_sel); end
        end
        fidx = fidx + 1;
    end
    pop = newpop;
    [pop, F] = non_dominated_sort(pop);
    pop = crowding_distance(pop, F);
end

% plot final Pareto (front 1)
FinalPareto = pop([pop.Rank]==1);
Obj = vertcat(FinalPareto.Objectives);
figure; plot(Obj(:,1), Obj(:,2),'r*'); grid on; xlabel('f_1'); ylabel('f_2');
title(['Pareto Front - ' problem_name]);

File: Python Implementation with fitness sharing/NSGA2_main.py
import numpy as np
import matplotlib.pyplot as plt
from individual import Individual
from get_problem_details import get_problem_details
from evaluate_objectives import evaluate_objectives
from non_dominated_sort import non_dominated_sort

# Import Hybrid Diversity Components
from crowding_distance import calculate_crowding_distance 
from fitness_share import calculate_fitness_sharing 
from tournament_selection import tournament_selection
from genetic_operators import genetic_operators

# ---------------- CONFIGURATION ----------------
# List of all problems to run
ALL_PROBLEMS = [
    'ZDT1', 'ZDT3', 'Kursawe', 'Fonseca', 
    'Schaffer1', 'Schaffer2', 'BinhKorn', 'ChankongHaimes'
]

N_POP = 50
MAX_IT = 100
PC = 0.9      # Crossover Prob
ETA_C = 15    # Crossover Distribution Index
ETA_M = 20    # Mutation Distribution Index

# ---------------- SINGLE-PROBLEM RUN SETUP ----------------

PROBLEM_NAME = ALL_PROBLEMS[6] # Select the problem to run

# --- START NEW FIGURE FOR THE PROBLEM ---
plt.figure(figsize=(7, 5)) 

print(f"\n--- Starting NSGA-II (Hybrid Diversity) for problem: {PROBLEM_NAME} ---")

# Get problem details
n_var, var_min, var_max = get_problem_details(PROBLEM_NAME)
PM = 1.0 / n_var

# Initialize Population
pop = []
for _ in range(N_POP):
    ind = Individual()
    ind.position = np.random.uniform(var_min, var_max)
    ind.objectives, ind.constraint_violation = evaluate_objectives(ind.position, PROBLEM_NAME)
    pop.append(ind)

# Initial Sort and Diversity Calculation (Calculate BOTH metrics)
pop, fronts = non_dominated_sort(pop)
pop, fronts = calculate_crowding_distance(pop, fronts)
pop = calculate_fitness_sharing(pop)

# ---------------- MAIN LOOP ----------------
for it in range(MAX_IT):
    
    # Create Offspring using the hybrid tournament selection
    offspring_pop = []
    while len(offspring_pop) < N_POP:
        p1 = tournament_selection(pop)
        p2 = tournament_selection(pop)
        
        c1_pos, c2_pos = genetic_operators(p1.position, p2.position, var_min, var_max, PC, ETA_C, PM, ETA_M)
        
        child1 = Individual()
        child1.position = c1_pos
        child1.objectives, child1.constraint_violation = evaluate_objectives(c1_pos, PROBLEM_NAME)
        offspring_pop.append(child1)
        
        if len(offspring_pop) < N_POP:
            child2 = Individual()
            child2.position = c2_pos
            child2.objectives, child2.constraint_violation = evaluate_objectives(c2_pos, PROBLEM_NAME)
            offspring_pop.append(child2)

    # Merge Populations
    combined_pop = pop + offspring_pop
    
    # Sort Combined Population
    combined_pop, fronts = non_dominated_sort(combined_pop)
    
    # Calculate BOTH metrics on the new population
    combined_pop, fronts = calculate_crowding_distance(combined_pop, fronts)
    combined_pop = calculate_fitness_sharing(combined_pop)
    
    # Survival Selection (Elitism)
    new_pop = []
    front_idx = 0
    
    while len(new_pop) < N_POP:
        front = fronts[front_idx]
        
        if len(new_pop) + len(front) <= N_POP:
            new_pop.extend(front)
        else:
            # --- HYBRID SURVIVAL SELECTION LOGIC for the boundary front ---
            
            # Check if this front is Rank 1 and it's the boundary front being trimmed
            if front_idx == 0: 
                # Use Crowding Distance (Higher CD is better for trimming the Pareto front)
                front.sort(key=lambda x: x.crowding_distance, reverse=True) 
            else:
                # Use Shared Fitness (Lower Niche Count is better for trimming deep fronts)
                front.sort(key=lambda x: x.shared_fitness, reverse=False) 

            needed = N_POP - len(new_pop)
            new_pop.extend(front[:needed])
            
        front_idx += 1
        
    pop = new_pop
    
    # Re-sort and recalculate for the next iteration
    pop, fronts = non_dominated_sort(pop)
    pop, fronts = calculate_crowding_distance(pop, fronts)
    pop = calculate_fitness_sharing(pop)

# ---------------- VISUALIZATION ----------------

pareto_front = [ind for ind in pop if ind.rank == 1]
f1_vals = [ind.objectives[0] for ind in pareto_front]
f2_vals = [ind.objectives[1] for ind in pareto_front]

# Plotting on the current figure context
plt.scatter(f1_vals, f2_vals, c='red', marker='o', s=10)
plt.title(f'Hybrid Front - {PROBLEM_NAME}', fontsize=12)
plt.xlabel(r'$f_1(\mathbf{x})$', fontsize=10)
plt.ylabel(r'$f_2(\mathbf{x})$', fontsize=10)
plt.grid(True, linestyle='--')

# Custom axes settings for ChankongHaimes
if PROBLEM_NAME == 'ChankongHaimes':
    plt.xlim(0, 250)
    plt.ylim(-250, 0)
    plt.xticks(np.arange(0, 251, 50))
    plt.yticks(np.arange(-250, 1, 50))

print(f"--- Optimization Finished for {PROBLEM_NAME} ---")

# --- SHOW THE RESULT ---
plt.tight_layout()
plt.show()